task6-hello-module-procfs/hello.c

```diff
@@ -0,0 +1,106 @@
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/moduleparam.h>
+#include <linux/printk.h>
+#include <linux/proc_fs.h>
+#include <linux/uaccess.h>
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Kriz");
+MODULE_DESCRIPTION("Hello_world module with simple canculator (sum, sub and mul)");
+MODULE_VERSION("0.1");
+
+#define PROC_BUFFER_SIZE 100
+
+static char   procfs_buffer[PROC_BUFFER_SIZE] = {0};
+static size_t procfs_buffer_size = 0;
+
+static struct proc_dir_entry *proc_file;
+static struct proc_dir_entry *proc_folder;
+
+#define PROC_FILE_NAME "dummy"
+#define PROC_DIR_NAME  "hello"
+
+static ssize_t hello_read(struct file *File, char __user* buffer, size_t count, loff_t *offset)
+{
+	ssize_t to_copy, not_copied, delta;
+	
+	if(0 == procfs_buffer_size)
+	{
+		return 0;
+	}
+	
+	to_copy = min(count, procfs_buffer_size);
+
+	not_copied = copy_to_user(buffer, procfs_buffer, to_copy);
+
+	delta = to_copy - not_copied;
+	procfs_buffer_size =- delta;
+
+	return delta;
+}
+
+/*static ssize_t hello_write(struct file *file, const char __user* buffer, size_t  count, loff_t *offset)
+{
+	ssize_t to_copy, not_copied, delta;
+	
+	to_copy = min(count, PROC_BUFFER_SIZE);
+	
+	not_copied = copy_from_user(procfs_buffer, buffer, to_copy);
+
+	delta = to_copy - not_copied;
+
+	procfs_buffer_size = delta;
+
+	return delta;
+}*/
+
+static struct proc_ops fops = {
+	.proc_read = hello_read
+	//.proc_write = hello_write
+};
+
+static int op1 = -1;
+module_param(op1, int, 00644);
+MODULE_PARM_DESC(op1, "First operator for calculation");
+
+static int op2 = -1;
+module_param(op2, int, 00644);
+MODULE_PARM_DESC(op2, "Second operator for calculation");
+
+static int __init hello_init(void) {
+	proc_folder = proc_mkdir(PROC_DIR_NAME, NULL);
+	if(!proc_folder)
+	{
+		pr_err("HELLO: Error: COULD not create /proc/%s/ folder\n", PROC_DIR_NAME);
+		return -ENOMEM;
+	}
+
+	proc_file = proc_create(PROC_FILE_NAME, 0666, proc_folder, &fops);
+	if(!proc_file)
+	{
+		pr_err("HELLO: Error: Could not initialize /proc/%s/%s\n", PROC_DIR_NAME, PROC_FILE_NAME);
+		proc_remove(proc_file);
```

Comment:
No sense to call `proc_remove(proc_file);` because `proc_file` is not valid (it contains an error code)